level2/derived.lisp

```diff
@@ -420,6 +420,33 @@ by an and pattern. Example: (plist :key1 _ :key2 value)"
                      `(property ,key ,pattern)))
                   (plist-alist args))))
 
+
+(define-condition hash-table-odd-number-of-entries-warning (simple-style-warning)
+  ())
+
+(define-condition hash-table-no-entries-warning (simple-style-warning)
+  ())
+
+(defpattern hash-table-entries (&rest keys-and-value-patterns)
+  "(HASH-TABLE-ENTRIES { KEY PATTERN }*)
+
+Matches hash-table which has KEY set to value matching PATTERN. Multiple KEY PATTERN pairs can be provided,
+e.g. (hash-table-entries :key1 _ :key2 value :key3 1)"
+  (cond ((oddp (length keys-and-value-patterns))
+         (warn 'hash-table-odd-number-of-entries-warning
+               :format-control "Odd number of arguments given to HASH-TABLE-ENTRIES pattern. Value for ~S will be interpreted as NIL."
+               :format-arguments (last keys-and-value-patterns)))
+        ((null keys-and-value-patterns)
+         (warn 'hash-table-no-entries-warning
+               :format-control "Empty HASH-TABLE-ENTRIES pattern is equivalent to (TYPE HASH-TABLE).")))
+  (flet ((key-value-pattern (key pattern)
+           (with-gensyms (it)
+             `(guard1 ,it (nth-value 1 (gethash ,key ,it))
+                      (gethash ,key ,it) ,pattern))))
```

Comment:
This part is computing gethash twice.
Unfortunately there is no way to directly express multiple values in guard1.
Perhaps you can combine `multiple-value-list` + `(guard1 (,it :dynamic-extent t) ... `

Reply:
I could not figure out how `:dynamic-extent` solves it... But this works:
```
(with-gensyms (it val-and-found-p)
  `(guard1 ,it t
           (multiple-value-list (gethash ,key ,it))
           (guard1 ,val-and-found-p (second ,val-and-found-p)
                   (first ,val-and-found-p) ,pattern)))
```
How does it look?

Reply:
yes, that is what I was thinking. Declaring `it` dynamic-extent will make the cons cell stack allocated.
Pattern matching ideally should not cause consing --- it should be just a powerful `cond`.

Reply:
oops, actually it is `val-and-found-p` that must be declared.

Reply:
hmmm.... not sure if it can actually take effect

Reply:
Added new code. I looked at output of `disassemble` to try to figure out if `:dynamic-extend t` changes anything and it seems that it does. On `sbcl` references to `thread.alloc-region` were gone afterwards, and on `ccl` `(jmp (@ .SPCONSLIST))` changed to `(jmp (@ .SPSTKCONSLIST))`. So I assume that's good news?

Reply:
Thats a good sign. Lastly, will you update the README entry and wiki page, then good to go

Reply:
Writing documentation I noticed that there is `property!` which led me to understand that `alist`, `plist` can match keys that don't exist. So for consistency sake, should I split `hash-table-entries` into two? `hash-table-entries` and `hash-table-entries!`?

Reply:
sounds like a good idea.

Reply:
So now there are 4 patterns `...-entry`, `...-entry!`, `...-entries`, and `...-entries!`. Not sure how to update wiki, have commit on fork here https://github.com/dop/trivia-wiki/commit/f42a2be3e1f7aa064159a8e7c077c55c05260336. Btw, mostly copied `property` implementation because noticed `sbcl` warnings of failure to stack allocate result of `multiple-value-list` in some cases.